DOC-558:  Custom Script Functions (#638)
Macrometacorp/docs: docs/cep/query-guide/custom-script-functions.md

---
sidebar_position: 80
title: Custom Script Functions
---

> Scripts allow you to write functions in JavaScript and execute them within queries for a stream. Functions defined with scripts can be accessed in queries similar to any other inbuilt function. You can use function definitions to define these scripts.

Reviewer: ```suggestion
Scripts enable you to write functions in JavaScript and execute them within queries for a stream. A query can access these custom functions the same way as pre-built functions. Use function definitions to define these scripts.
```

Not sure that the second sentence is really necessary

> Function parameters are passed into the function logic as an `Object[]` and with the name `data`.

Reviewer: ```suggestion
Function parameters are considered an `Object[]` with the name `data`.
```

Do we need to specify `Object[]` syntax instead of just saying "considered an object" or "considered objects"

## Purpose

> Scripts allow you to define a function operation that is not provided in stream worker built-in functions. You do not need to write an extension to define the function logic.

Reviewer: ```suggestion
Scripts enable you to define a function that is not built into stream workers. You do not need to write an extension to define the function logic.
```

## Syntax

The syntax for a script definition is as follows.

```js
define function <function name>[<javascript>] return <return type> {
    <operation of the function>
};
```

> The defined function can be used in the queries similar to inbuilt functions as follows.

Reviewer: ```suggestion
The defined function can be used in the queries similar to in-built functions as follows.
```

```js
<function name>( (<function parameter>(, <function parameter>)*)? )
```

> The custom functions defined with these function definitions have higher precedence compared to other functions.

Reviewer: ```suggestion
Custom functions have higher precedence than in-built functions.
```

## Parameters

The following parameters are configured when defining a script.

| Parameter     | Description |
| ------------- |-------------|

> | function name | 	The name of the function. By convention, `camelCase` is used for the function name.|

Reviewer: ```suggestion
| function name | 	The name of the function. Standard convention is `camelCase`.|
```

> | language name | The name of the programming language used to define the script. Currently, Macrometa supports `javascript`.|

Reviewer: ```suggestion
| language name | The name of the programming language used to define the script. Macrometa supports `javascript`.|
```

Should we change this to say "`javascript` and `python`"?  Or do we only support JS in this particular instance?

Author: We only support JavaScript for this.

| return type | The attribute type of the function’s return. This can be `int`, `long`, `float`, `double`, `string`, `bool`, or `object`. Here the function implementer should be responsible for returning the output attribute on the defined return type for proper functionality.

> | operation of the function| The execution logic of the function. This logic should be written in the language specified under the `language name`, and it should return the output in the data type specified in the `return type` parameter.

Reviewer: ```suggestion
| operation of the function| The running logic of the function. This logic should be written in the language specified under the `language name`, and it should return the output in the data type specified in the `return type` parameter.
```
